core/src/impl/Kokkos_FunctorAnalysis.hpp

```diff
@@ -398,11 +398,15 @@ struct FunctorAnalysis {
   //----------------------------------------
   // parallel_reduce join operator
 
-  template <class F, bool is_array = candidate_is_array>
+  template <class F, bool is_volatile, bool is_array = candidate_is_array>
   struct has_join_no_tag_function;
 
+#define KOKKOS_IMPL_LOVATILE  // detect join member functions with
+                              // volatile-qualified arguments for backward
+                              // compatibility
```

Comment:
A more descriptive name would avoid the necessity for a clarifying comment.
```suggestion
#define KOKKOS_IMPL_DETECT_VOLATILE
```

Reply:
Ultimately, it would either be unconditional, or would be a deprecation flag, I think. Having the methods take `volatile` has a performance cost that we should steer client code away from.